### PR TITLE
Move dof table functionality

### PR DESCRIPTION
**Pull request by TomFischer**
As titled.

### CODE REVIEW

**Comment by chleh** [comment]
Please have a look at [DOFTableUtil](/ufz/ogs/tree/master/NumLib/DOF/DOFTableUtil.cpp). There is already similar code.

**Comment by TomFischer** [comment]
> Please have a look at DOFTableUtil. There is already similar code.

Thanks for the hint, I'll check this tomorrow.

File: NumLib/DOF/DOFTableUtil.cpp [hunk 1]
```diff
@@ -44,6 +44,42 @@ std::vector<GlobalIndexType> getIndices(
     return indices;
 }
 
+NumLib::LocalToGlobalIndexMap::RowColumnIndices getRowColumnIndices(
```
Comment (norihiro-w) [inline NumLib/DOF/DOFTableUtil.cpp]: It looks for me the function is similar to `getIndices()`. Can't you reuse it?

File: NumLib/DOF/DOFTableUtil.cpp [hunk 2]
```diff
@@ -44,6 +44,42 @@ std::vector<GlobalIndexType> getIndices(
...
+    {
+        auto const& idcs = dof_table(id, c).rows;
+        indices.reserve(indices.size() + idcs.size());
+        indices.insert(indices.end(), idcs.begin(), idcs.end());
+    }
+
+    return NumLib::LocalToGlobalIndexMap::RowColumnIndices(indices, indices);
+}
+
+void getVectorValues(
```
Comment (norihiro-w) [inline NumLib/DOF/DOFTableUtil.cpp]: `GlobalVector` such  as `EigenVector` has `get(std::vector<IndexType> const& indices)`. maybe you can use it.

File: NumLib/DOF/DOFTableUtil.cpp (hunk 1 above)
Comment (TomFischer) [inline NumLib/DOF/DOFTableUtil.cpp]: Indeed these functions are very similar. The important difference is that the `getRowColumnIndices` does not create the result vector `indices` again and again for every call of the function. Anyway, in this PR I only want to move existing code such that I can use it a second time at another place.

File: NumLib/DOF/DOFTableUtil.cpp (hunk 2 above)
Comment (TomFischer) [inline NumLib/DOF/DOFTableUtil.cpp]: Yes, you are right. The function `get()` is already used internally in `getVectorValues()`, see line 78. The function is a wrapper for the `GlobalVector::get`  that ensures that the vector `local_x` has the appropriate size.

Comment (norihiro-w) [inline NumLib/DOF/DOFTableUtil.cpp]: is `get()` at line 78 `get(int)` or `get(std::vector<int>&)`?

File: NumLib/DOF/DOFTableUtil.cpp (hunk 1 above)
Comment (norihiro-w) [inline NumLib/DOF/DOFTableUtil.cpp]: i see. thanks for the explanation

File: NumLib/DOF/DOFTableUtil.cpp (hunk 2 above)
Comment (TomFischer) [inline NumLib/DOF/DOFTableUtil.cpp]: If I am not wrong it is `IDX_TYPE` which is probably `int`.

Comment (norihiro-w) [inline NumLib/DOF/DOFTableUtil.cpp]: what I actually wanted to say is that you can write it as below. maybe you already know it and intentionally you are not using it. 

```
local_x = x.get(r_c_indices.rows);
```

Comment (TomFischer) [inline NumLib/DOF/DOFTableUtil.cpp]: Thanks for the hint, I did not know this. I needed only the functionality in my own balance implementation. So I moved the code to a general accessible location.

**Comment by TomFischer** [comment]
@chleh Are the changes in TES okay?

**Comment by chleh** [comment]
Changes in TES are OK. Code looks good. So you might merge this PR.

I only want to make a short remark:
The functionality added is, as you said, that `getRowColumnIndices()` doesn't create the indices vector. However `x.get(r_c_indices.rows)` will always create a vector.
So if this optimization is really important, then it might be important in both places (note: IMHO it's currently not so important). Furthermore, if it's important, it might be important for `getIndices()`, too. And when changing `getIndices()`, one would have to check what that means for the usability of this function, i.e., whether the user has to write more code.
So this PR duplicates some functionality; I am not really happy with it. But in order to not block you: Merge and forget :smile:

**Comment by TomFischer** [comment]
I am unsure. Should I let `getVectorValues()` in the TES process?

**Comment by chleh** [comment]
For TES you can keep it as you have it in this PR.

**Pull request merged**